File: app/src/main/java/com/sppe/shrimppaste/adapter/GirlItemDecoration.java
package com.sppe.shrimppaste.adapter;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;

import com.sppe.shrimppaste.R;

/**
 * Created by WHF on 2017/11/5.
 */

public class GirlItemDecoration extends RecyclerView.ItemDecoration {

    private int leftMargin;
    private int rightMargin;
    private int bottomMargin;

    private Paint paint;

    public GirlItemDecoration(Context context) {
        leftMargin = context.getResources().getDimensionPixelSize(R.dimen.girl_item_left_margin);
        rightMargin = context.getResources().getDimensionPixelSize(R.dimen.girl_item_right_margin);
        bottomMargin = context.getResources().getDimensionPixelSize(R.dimen.girl_item_bottom_margin);

        paint = new Paint();
        paint.setColor(context.getResources().getColor(R.color.divide));
        paint.setStyle(Paint.Style.FILL);
        paint.setAntiAlias(true);
    }

    @Override
    public void onDraw(Canvas c, RecyclerView parent, RecyclerView.State state) {
        super.onDraw(c, parent, state);
        int count = parent.getChildCount();
        int index = 0;

        //仅有Footer
        if (count == 1) {
            return;
        }

        while (index < count) {
            View view = parent.getChildAt(index);
            float left = view.getLeft();
            float top = view.getTop();
            float right = view.getRight();
            float bottom = view.getBottom();

            //下
            c.drawRect(left - leftMargin, bottom, right + rightMargin, bottom + bottomMargin, paint);
            if (index % 2 == 0) {
                //右
                c.drawRect(right, top, right + rightMargin, bottom, paint);
            } else {
                //左
                c.drawRect(left - leftMargin, top, left, bottom, paint);
            }
            ++index;
        }
    }

    @Override
    public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state) {
        super.getItemOffsets(outRect, view, parent, state);
        int index = parent.getChildAdapterPosition(view);
        Log.e("====",""+index);
        if (index % 2 == 0) {
            //右
            outRect.set(0, 0, rightMargin, bottomMargin);
        } else {
            //左
            outRect.set(leftMargin, 0, 0, bottomMargin);
        }
    }
}

File: app/src/main/java/com/sppe/shrimppaste/adapter/GirlItemDecoration.kt
package com.sppe.shrimppaste.adapter

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.support.v7.widget.RecyclerView
import android.view.View
import com.sppe.shrimppaste.R

/**
 *
 * Created by @author WangHaoFei on 2017/11/2.
 */
class GirlItemDecoration(val context: Context) : RecyclerView.ItemDecoration() {

    private val leftMargin = context.resources.getDimensionPixelSize(R.dimen.girl_item_left_margin)
    private val rightMargin = context.resources.getDimensionPixelSize(R.dimen.girl_item_right_margin)
    private val bottomMargin = context.resources.getDimensionPixelSize(R.dimen.girl_item_bottom_margin)

    private val paint = Paint()

    init {
        paint.color = context.resources.getColor(R.color.divide)
        paint.style = Paint.Style.FILL
        paint.isAntiAlias = true
    }

    override fun onDraw(c: Canvas?, parent: RecyclerView?, state: RecyclerView.State?) {
        super.onDraw(c, parent, state)
        val count = parent!!.childCount
        var index = 0
        while (index < count) {
            val view = parent.getChildAt(index)
            val left = view.left.toFloat()
            val top = view.top.toFloat()
            val right = view.right.toFloat()
            val bottom = view.bottom.toFloat()

            c!!.drawRect(left - leftMargin, top, left, bottom, paint)
            c.drawRect(right, top, right + rightMargin, bottom, paint)
            c.drawRect(left - leftMargin, bottom, right + rightMargin, bottom + bottomMargin, paint)

            ++index
        }

    }

    override fun getItemOffsets(outRect: Rect?, view: View?, parent: RecyclerView?, state: RecyclerView.State?) {
        super.getItemOffsets(outRect, view, parent, state)
        outRect!!.set(leftMargin, 0, rightMargin, bottomMargin)
    }
}